[RISC-V] Allow intrinsics to be used with any pointer type.

hvdijk opened the pull request:
RISC-V does not use address spaces and leaves them available for user code to make use of. Intrinsics, however, required pointer types to use the default address space, complicating handling during lowering to handle non-default address spaces. When the intrinsics are overloaded, this is handled without extra effort.

This commit does not yet update Clang builtin functions to also permit pointers to non-default address spaces.

hvdijk commented:
At the moment this leaves a bunch of LLVM tests using the wrong signature for builtins, which is silently accepted. I will update those tests and update this PR once I have some way of verifying that the changes are complete.

wangpc-pp commented:
I don't know how many users are requiring this change, but I do know some people are developing RVV-based heterogeneous hardware/software. I am happy to see this and we should make more people be aware of this change, so I suggest to post it on Discourse as well.

wangpc-pp commented on clang/include/clang/Basic/riscv_vector.td:
```diff
@@ -63,11 +63,11 @@ multiclass RVVVLEFFBuiltin<list<string> types> {
             Ops.insert(Ops.begin(), llvm::PoisonValue::get(ResultType));
           Ops.push_back(ConstantInt::get(Ops.back()->getType(), PolicyAttrs));
-          IntrinsicTypes = {ResultType, Ops[4]->getType()};
+          IntrinsicTypes = {ResultType, Ops[4]->getType(), Ops[2]->getType()};
```
Should we make a convention here that the order of type should follow the argument order?

hvdijk replied:
The types in `IntrinsicTypes` have to be whatever order the intrinsic expects, which is result first, in order if there are multiple results, then each parameter in order. The fact that the VL type appears in the intrinsic's result requires it to appear in `IntrinsicTypes` ahead of the pointer type.

wangpc-pp reviewed (commented):
This PR is too huge and it almost makes my Chrome broken. :-(

hvdijk commented:
> I think [Halide will emit RISC-V LLVM IR intrinsics](https://github.com/halide/Halide/blob/bf9c55dd1392b87cfc82371ee40157786eb4ec78/src/CodeGen_RISCV.cpp#L357-L366), so it would be good to double check that the code it emits still works fine with these new intrinsics.

It may not, but this is not a change I'm planning on asking to be backported to release branches, and this is the sort of change that we wouldn't normally expect to remain stable across releases. Certainly Halide have no problem using `#if` to do things differently for different LLVM versions, https://github.com/halide/Halide/pull/8591/files. I'll check what it does when I can, though, specifically that *if* it doesn't work, it at least generates a sensible error that makes it easy to see how to update the code.

hvdijk committed:
[LLVM][RISC-V] Allow intrinsics to be used with any pointer type.

RISC-V does not use address spaces and leaves them available for user
code to make use of. Intrinsics, however, required pointer types to use
the default address space, complicating handling during lowering to
handle non-default address spaces. When the intrinsics are overloaded,
this is handled without extra effort.

This commit does not yet update Clang builtin functions to also permit
pointers to non-default address spaces.

hvdijk commented:
Rebased since #139698 was merged; the change in the newly added `addrspace.ll` test shows where this PR makes a difference (thanks for adding that!)

I've checked Halide now too. It is not affected by this because it does not generate any RISC-V intrinsics taking pointers, and only intrinsics taking pointers are changed by this PR.

4vtomat reviewed (approved):
LGTM, thanks~

wangpc-pp reviewed (approved):
> Thanks! @wangpc-pp Are you still reviewing as well? If you are, I'll hold off on merging in case you have comments.

LGTM! I always support this change, just need more eyes in case we may have some typo.

hvdijk commented:
> LGTM! I always support this change, just need more eyes in case we may have some typo.

Yep, it's a lot of files so it is easy for something to be missed, it helps that the majority of the files are tests where the risk of typos is low because we see the tests passing. Thanks, I'll wait till end of tomorrow then to see if anyone else wants to review, otherwise I think we can merge it.
